docs/architecture/deep-dives/state/index.md

```diff
@@ -0,0 +1,379 @@
+
+`KeyDefinition` builds on the idea of [`StateDefinition`](#statedefinition) but it gets more
+specific about the data to be stored. `KeyDefinition`s can also be instantiated in your own teams
+code. This might mean creating it in the same file as the service you plan to consume it or you may
+want to have a single `key-definitions.ts` file that contains all the entries for your team. Some
+example instantiations are:
+
+```typescript
+const MY_DOMAIN_DATA = new KeyDefinition<MyState>(MY_DOMAIN_DISK, "data", {
+  deserializer: (jsonData) => null, // convert to your data from json
+});
+
+// Or if your state is an array, use the built in helper
+const MY_DOMAIN_DATA: KeyDefinition<MyStateElement[]> = KeyDefinition.array<MyStateElement>(
+  MY_DOMAIN_DISK,
+  "data",
+  {
+    deserializer: (jsonDataElement) => null, // provide a deserializer just for the element of the array
+  },
+);
+
+// record
+const MY_DOMAIN_DATA: KeyDefinition<Record<string, MyStateElement>> =
+  KeyDefinition.record<MyStateValue>(MY_DOMAIN_DISK, "data", {
+    deserializer: (jsonDataValue) => null, // provide a deserializer just for the value in each key-value pair
+  });
+```
+
+The first argument to `KeyDefinition` is always the `StateDefinition` that this key should belong
+to. The second argument should be a human readable, camelCase formatted name of the `KeyDefinition`.
+For example, the accounts service may wish to store a known accounts array on disk and choose
+`knownAccounts` to be the second argument. This name should be unique amongst all other
+`KeyDefinition`s that consume the same `StateDefinition`. The responsibility of this uniqueness is
+on the team. As such, you should never consume the `StateDefinition` of another team in your own
+`KeyDefinition`. The third argument is an object of type
+[`KeyDefinitionOptions`](#keydefinitionoptions).
+
+##### `KeyDefinitionOptions`
+
+`deserializer` (required) - Takes a method that gives you your state in it's JSON format and makes
+you responsible for converting that into JSON back into a full JavaScript object, if you choose to
+use a class to represent your state that means having it's prototype and any method you declare on
+it. If your state is a simple value like `string`, `boolean`, `number`, or arrays of those values,
+your deserializer can be as simple as `data => data`. But, if your data has something like `Date`,
+which gets serialized as a string you will need to convert that back into a `Date` like:
+`data => new Date(data)`.
+
+`cleanupDelayMs` (optional) - Takes a number of milliseconds to wait before cleaning up the state
+after the last subscriber has unsubscribed. Defaults to 1000ms.
+
+### `StateProvider`
+
+`StateProvider` is an injectable service that includes 3 methods for getting state. These three
+methods are helpers for invoking their more modular siblings `ActiveStateProvider.get`,
+`SingleUserStateProvider.get`, and `GlobalStateProvider.get` and they have the following type
```

Comment:
❓ Is it ok to use the more tightly scoped siblings? Like, if I only care about the active user, can I just inject `ActiveUserStateProvider`?

Reply:
You can, the siblings are all injectable through DI or available right next to `StateProvider` (browser background or CLI). These providers all have a caching element to them where if state is requested for a `KeyDefinition` more than once they will make it the first time and return the same instance for any subsequent requests. But that cache is inside the siblings themselves so you won't even be skirting that by using `ActiveUserStateProvider` directly.

Reply:
🎨 Sweet! I might make it more explicit that if you only need one kind of state, that you should inject the smaller interface.

Reply:
I tried to make it more explicit as an option but I didn't go as far as to say they _should_ do it that way. It has a con where they will need to inject another service if they now have to go change the code that instantiates that service a minimum of 4 other places. Heaviness-wise the heaviest service is going to be `ActiveUserStateProvider` which is likely to be the most commonly used service. All the other providers take the same dependencies minus `AccountService`. So if they can use `ActiveUserStateProvider` they will get no new circular dependencies possibilities from just taking the `StateProvider` helper.